### PR TITLE
New component observer/cfgardenobserver

### PR DESCRIPTION
**Pull request by CemDK**
**Description:** 
We would like to implement a new observer for [Cloudfoundry](https://www.cloudfoundry.org/) containers/applications. The idea is not make use of the main API but the local one, which is available as unix socket on each node and manages the containers lifecycle. The main API would remain as optional and only to get Application info (which only involves one GET http request once the app id is known).

**Link to tracking Issue:** [33618](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33618)

**Testing:** First component PR

**Documentation:** Added Readme

### CODE REVIEW

File: extension/observer/cfgardenobserver/testdata/config.yaml
Comment (crobert-1) [inline extension/observer/cfgardenobserver/testdata/config.yaml:4]:
```suggestion
  endpoint: http://localhost:8080
```
I assume `b` is a typo?

File: extension/observer/cfgardenobserver/metadata.yaml
Comment (crobert-1) [inline extension/observer/cfgardenobserver/metadata.yaml:11]:
```suggestion
```

File: extension/observer/cfgardenobserver/metadata.yaml
Comment (crobert-1) [inline extension/observer/cfgardenobserver/metadata.yaml:9]: Please add all code owners here, as well as [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/githubgen/allowlist.txt).

File: extension/observer/cfgardenobserver/extension_test.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension_test.go:35]: Can we use a shorter timeout here? I don't want to introduce flakiness, but 2 seconds seems longer than necessary.

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:14]:
```suggestion
Cloudfoundry is a platform as a service (PaaS) which has implemented their own container technology. The component in charge of providing a local API in each node (also known as "diego-cell") is named Garden.
```

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:19]: It would be good to include a full configuration example here. We should also make it clear which variables are required and which are optional.

Config options are usually in a list, rather than sections. It may be good to convert these into a list to be more uniform with other READMEs.

File: extension/observer/cfgardenobserver/config_test.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config_test.go:33]: I suspect this is an improper type setting, where the number here needs to be a duration, rather than a plain int.

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:34]: This method isn't defined anywhere. This PR is already a bit big as the skeleton, so I'd prefer removing the reference rather than adding the method.

**Comment by tomasmota** [comment]
Thanks for the review @crobert-1 , we have made the requested changes.

**Comment by tomasmota** [comment]
sorry to any maintainers that got pinged, messed up a rebase

File: extension/observer/cfgardenobserver/config.go
Comment (MovieStoreGuy) [inline extension/observer/cfgardenobserver/config.go:7]: I suspect that comment is a side affect of formatter tripping up.

Do you mind fixing up the placement?

File: extension/observer/cfgardenobserver/extension.go
Comment (MovieStoreGuy) [inline extension/observer/cfgardenobserver/extension.go:29]: If you want to make it easier on yourself if you don't need to implement `Start` and `Shutdown`, you can embed `component.StartFunc` and `component.ShutdownFunc` within the main type and it will implement safe methods for external methods to call.

File: extension/observer/cfgardenobserver/extension_test.go
Comment (MovieStoreGuy) [inline extension/observer/cfgardenobserver/extension_test.go:35]: Can I kindly ask that you don't introduce a timed delay within testing since it often encourages anti patterns on how handle start up and shutdown.

File: extension/observer/cfgardenobserver/factory.go
Comment (MovieStoreGuy) [inline extension/observer/cfgardenobserver/factory.go:17]:
```suggestion
	defaultCollectionInterval = 1 * time.Minute
```

**Comment by MovieStoreGuy** [comment]
> sorry to any maintainers that got pinged, messed up a rebase

No dramas, happens from time to time :)

**Comment by crobert-1** [comment]
_There are quite a few CI/CD action failures, and they appear to be related to the changes in this PR. I'll hold off on reviewing again until they're resolved._

**Comment by tomasmota** [comment]
Thanks @crobert-1 and @MovieStoreGuy, I have made the changes and the ci is now passing. Thank you for your reviews

File: cmd/otelcontribcol/builder-config.yaml
Comment (crobert-1) [inline cmd/otelcontribcol/builder-config.yaml:53]: Nit: Could this line be moved up to be with the other observer extensions, placing it in alphabetical order?

File: .github/CODEOWNERS
Comment (crobert-1) [inline .github/CODEOWNERS:110]: Nit: Could this line be moved to be in alphabetical order? (I assume this was added manually, but let me know if this was generated automatically and placed out of order)

Also, I need to be added to this list as well.

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:1]: Nit: We may want to expand Cloud Foundry here instead of just using the initials, to make it more clear what we're referring to.

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:14]:
```suggestion
Cloud Foundry (CF) is a platform as a service (PaaS) which has implemented their own container technology. The component in charge of providing a local API in each node (also known as "diego-cell") is named Garden.
```

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:23]:
```suggestion
    # url of the garden socket, defaults to unix:///var/vcap/data/garden/garden.sock 
```

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:16]: From what I can tell, `Garden` should be capitalized in all documentation and comments. (Code references can remain lower case.)

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:52]: The description here isn't immediately clear, can you clarify? Is this simply saying the value for the  exposed `type` key will always be `container`?

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config.go:15]:
```suggestion
	// RefreshInterval determines the frequency at which the observer
```

File: extension/observer/cfgardenobserver/factory_test.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/factory_test.go:20]:
```suggestion
func TestCreateCFGardenObserver(t *testing.T) {
```

File: versions.yaml
Comment (crobert-1) [inline versions.yaml:78]: Can this be moved to be in alphabetical order?

File: extension/observer/cfgardenobserver/README.md
Comment (tomasmota) [reply extension/observer/cfgardenobserver/README.md:52]: Yes, it will always be `container`. We just did the same as the host observer here, do you want me to change both to be clearer or none of them? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/hostobserver/README.md?plain=1#L35

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config.go:10]:
```suggestion
// Config defines configuration for CF Garden observer.
```

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config.go:12]:
```suggestion
	// The URL of the CF Garden api.  Default is "unix:///var/vcap/data/garden/garden.sock"
```

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [reply extension/observer/cfgardenobserver/README.md]: Thanks for the reference. My personal preference is to lean on the side of extra information, so I like having the extra description here.

**Comment by jriguera** [comment]
Hey @atoulme @dmitryax can you review this PR? 
Thanks!

**Review by crobert-1 — APPROVED** [review]

File: extension/observer/cfgardenobserver/config.go
Comment (atoulme) [inline extension/observer/cfgardenobserver/config.go:13]: Is the endpoint always a path, or can it be a URL with the http/s scheme? asking as this lib might come handy: https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/confignet

Note this can be done later in a subsequent PR

**Review by atoulme — APPROVED** [review]

File: extension/observer/cfgardenobserver/config.go
Comment (jriguera) [reply extension/observer/cfgardenobserver/config.go:13]: In CF is always a unix socket, but in terms of Garden it can be a http url. We will use this library! Thanks.

**Pull request merged**
